src/shim.c

```diff
@@ -954,6 +958,52 @@ connect_to_proxy(struct cc_shim *shim)
 	return false;
 }
 
+inline void sleep_ms(int ms)
+{
+	struct timespec ts = { 0, ms * 1000000L };
+	nanosleep(&ts, NULL);
+}
+
+/*
+ * Try to re-establish tcp/unix connection with proxy in RECONNECT_TIMEOUT_S
+ * seconds.
+ *
+ * \param shim \ref cc_shim
+ *
+ * \return true on success, false on failure
+ */
+bool reconnect_to_proxy(struct cc_shim *shim)
+{
+	// How many seconds shim tries to reconnect to proxy
+	const int RECONNECT_TIMEOUT_S = 10;
+	// Period between each attempt in milliseconds
+	const int RECONNECT_POLL_MS = 333;
+
+	shim_warning("Reconnecting to proxy\n");
+	int time = 0;
+
+try_again:
+	sleep_ms(RECONNECT_POLL_MS);
+	time += RECONNECT_POLL_MS;
+	if (time >= RECONNECT_TIMEOUT_S * 1000) {
+		shim_error("Failed to reconnect to proxy in %d sec\n",
+				RECONNECT_TIMEOUT_S);
+		return false;
+	}
+	if (! connect_to_proxy(shim)) {
+		goto try_again;
+	}
+
+	shim_debug("Sending connect command\n");
+	if (! send_connect_command(shim)) {
+		shim_error("Could not send connect command to proxy\n");
+		goto try_again;
+	}
+	add_pollfd(poll_fds, PROXY_SOCK_INDEX, shim->proxy_sock_fd,
```

Comment:
Could you add a comment explaining why this is being called again for `PROXY_SOCK_INDEX`.